File: src/AIMP.SDK/MusicLibrary/DataStorage/AimpDataField.cs
// ----------------------------------------------------
// 
// AIMP DotNet SDK
// 
// Copyright (c) 2014 - 2020 Evgeniy Bogdan
// https://github.com/martin211/aimp_dotnet
// 
// Mail: mail4evgeniy@gmail.com
// 
// ----------------------------------------------------

namespace AIMP.SDK.MusicLibrary.DataStorage
{
    /// <summary>
    /// Class AimpDataField.
    /// Implements the <see cref="AIMP.SDK.MusicLibrary.DataStorage.IAimpDataField" />
    /// </summary>
    /// <seealso cref="AIMP.SDK.MusicLibrary.DataStorage.IAimpDataField" />
    public class AimpDataField : IAimpDataField
    {
        /// <summary>
        /// The aimpml reserved field identifier
        /// </summary>
        public static readonly string AIMPML_RESERVED_FIELD_ID = "ID";

        /// <summary>
        /// The aimpml reserved field filename
        /// </summary>
        public static readonly string AIMPML_RESERVED_FIELD_FILENAME = "FileName"; // !REQUIRED! string

        /// <summary>
        /// The aimpml reserved field filesize
        /// </summary>
        public static readonly string AIMPML_RESERVED_FIELD_FILESIZE = "FileSize"; // Int64, in bytes

        /// <summary>
        /// The aimpml reserved field duration
        /// </summary>
        public static readonly string AIMPML_RESERVED_FIELD_DURATION = "Duration"; // double, in seconds

        /// <summary>
        /// The aimpml reserved field usermark
        /// </summary>
        public static readonly string AIMPML_RESERVED_FIELD_USERMARK = "UserMark"; // double, 0.0 .. 5.0

        /// <summary>
        /// Initializes a new instance of the <see cref="AimpDataField" /> class.
        /// </summary>
        /// <param name="name">The name.</param>
        /// <param name="type">The type.</param>
        /// <param name="flags">The flags.</param>
        public AimpDataField(string name, AimpDataFieldType type,
            AimpDataFieldFlagsType flags = AimpDataFieldFlagsType.None)
        {
            Name = name;
            Type = type;
            Flags = flags;
        }

        /// <summary>
        /// Gets or sets the user data. That data will be attached to the instance of interface.
        /// </summary>
        /// <value>The custom.</value>
        public string Custom { get; set; }

        /// <summary>
        /// Gets or sets the name.
        /// </summary>
        /// <value>The name.</value>
        public string Name { get; set; }

        /// <summary>
        /// Gets or sets the type of field <see cref="AimpDataFieldType" />.
        /// </summary>
        /// <value>The type.</value>
        public AimpDataFieldType Type { get; set; }

        /// <summary>
        /// Gets or sets the data field flags.
        /// </summary>
        /// <value>The flags.</value>
        public AimpDataFieldFlagsType Flags { get; set; }

        /// <summary>
        /// Gets or sets the field image.
        /// </summary>
        /// <value>The image.</value>
        public ImageType Image { get; set; }

        /// <summary>
        /// Gets or sets the display value.
        /// </summary>
        /// <value>The display value.</value>
        public string DisplayValue { get; set; }
    }
}

File: src/aimp_dotnet/SDK/MusicLibrary/DataStorage/AimpDataField.cpp
// ----------------------------------------------------
// AIMP DotNet SDK
// Copyright (c) 2014 - 2020 Evgeniy Bogdan
// https://github.com/martin211/aimp_dotnet
// Mail: mail4evgeniy@gmail.com
// ----------------------------------------------------

#include "Stdafx.h"
#include "AimpDataField.h"

using namespace AIMP::SDK;

AIMP::SDK::AimpDataField::AimpDataField(IAIMPMLDataField* aimpObject) : AimpObject(aimpObject)
{
}

String^ AIMP::SDK::AimpDataField::Custom::get()
{
    return PropertyListExtension::GetString(InternalAimpObject, AIMPML_FIELD_PROPID_CUSTOM);
}

void AIMP::SDK::AimpDataField::Custom::set(String^ value)
{
    PropertyListExtension::SetString(InternalAimpObject, AIMPML_FIELD_PROPID_CUSTOM, value);
}

String^ AIMP::SDK::AimpDataField::Name::get()
{
    return PropertyListExtension::GetString(InternalAimpObject, AIMPML_FIELD_PROPID_NAME);
}

void AIMP::SDK::AimpDataField::Name::set(String^ value)
{
    PropertyListExtension::SetString(InternalAimpObject, AIMPML_FIELD_PROPID_NAME, value);
}

AimpDataFieldType AIMP::SDK::AimpDataField::Type::get()
{
    return static_cast<AimpDataFieldType>(PropertyListExtension::GetInt32(InternalAimpObject, AIMPML_FIELD_PROPID_TYPE)
    );
}

void AIMP::SDK::AimpDataField::Type::set(AimpDataFieldType value)
{
    PropertyListExtension::SetInt32(InternalAimpObject, AIMPML_FIELD_PROPID_TYPE, int(value));
}

AimpDataFieldFlagsType AIMP::SDK::AimpDataField::Flags::get()
{
    return static_cast<AimpDataFieldFlagsType>(PropertyListExtension::GetInt32(
        InternalAimpObject, AIMPML_FIELD_PROPID_FLAGS));
}

void AIMP::SDK::AimpDataField::Flags::set(AimpDataFieldFlagsType value)
{
    PropertyListExtension::SetInt32(InternalAimpObject, AIMPML_FIELD_PROPID_FLAGS, int(value));
}

ImageType AIMP::SDK::AimpDataField::Image::get()
{
    return static_cast<ImageType>(PropertyListExtension::GetInt32(InternalAimpObject, AIMPML_FIELD_PROPID_IMAGE));
}

void AIMP::SDK::AimpDataField::Image::set(ImageType value)
{
    PropertyListExtension::SetInt32(InternalAimpObject, AIMPML_FIELD_PROPID_IMAGE, int(value));
}

String^ AIMP::SDK::AimpDataField::DisplayValue::get()
{
    return PropertyListExtension::GetString(InternalAimpObject, AIMPML_FIELD_PROPID_DISPLAYVALUE);
}

void AIMP::SDK::AimpDataField::DisplayValue::set(String^ value)
{
    PropertyListExtension::SetString(InternalAimpObject, AIMPML_FIELD_PROPID_DISPLAYVALUE, value);
}
